api/paidAction/zap.js

```diff
@@ -0,0 +1,169 @@
+    acts = await tx.itemAct.findMany({ where: { id: { in: actIds } }, include: { item: true } })
+  } else {
+    throw new Error('No invoice or actIds')
+  }
+
+  const msats = acts.reduce((a, b) => a + BigInt(b.msats), BigInt(0))
+  const sats = msats / BigInt(1000)
+  const itemAct = acts.find(act => act.act === 'TIP')
+
+  // give user and all forwards the sats
+  await tx.$executeRaw`
+    WITH forwardees AS (
+      SELECT "userId", ${itemAct.msats} * pct / 100 AS msats
+      FROM "ItemForward"
+      WHERE "itemId" = ${itemAct.itemId}
+    ), total_forwarded AS (
+      SELECT COALESCE(SUM(msats), 0) as msats
+      FROM forwardees
+    ), forward AS (
+      UPDATE users
+      SET msats = users.msats + forwardees.msats
+      FROM forwardees
+      WHERE users.id = forwardees."userId"
+    )
+    UPDATE users
+    SET msats = msats + ${itemAct.msats} - (SELECT msats FROM total_forwarded)
+    WHERE id = ${itemAct.item.userId}`
+
+  // perform denomormalized aggregates: weighted votes, upvotes, msats, lastZapAt, bountyPaidTo
+  // XXX this is vulnerable to serialization anomalies in the case of multiple zaps from the same user
+  // on the same item at the exact same time
+  // basically, in read committed, the select subqueries will see a different snapshot than update if
+  // the update conflicts with another update and new rows are inserted while the update is blocked
+  // e.g. it's possible that our aggregates in the zapped CTE will be stale
+  // ... other than changing the isolation level, we could:
+  // 1. store the aggregate zap from a user on an item in a separate table that can be locked `FOR UPDATE`
+  // 2. store the aggregate zap from a user in a jsonb column on the item
+  // see: https://stackoverflow.com/questions/61781595/postgres-read-commited-doesnt-re-read-updated-row?noredirect=1#comment109279507_61781595
+  // or: https://www.cybertec-postgresql.com/en/transaction-anomalies-with-select-for-update/
+  await tx.$executeRaw`
+    WITH zapper AS (
+      SELECT * FROM users WHERE id = ${itemAct.userId}
+    ), zapped AS (
+      SELECT COALESCE(SUM("ItemAct".msats) / 1000, 0)  as sats
+      FROM "ItemAct"
+      WHERE "ItemAct"."userId" = ${itemAct.userId}
+      AND "ItemAct"."itemId" = ${itemAct.itemId}
+      AND NOT "ItemAct".id = ANY (${actIds})
+      AND act IN ('TIP', 'FEE')
+      AND ("ItemAct"."invoiceActionState" IS NULL OR "ItemAct"."invoiceActionState" = 'PAID')
+    ),  zap AS (
+      SELECT (zapper.trust *
+        CASE WHEN zapped.sats = 0
+          THEN LOG(${sats})
+          ELSE LOG((zapped.sats + ${sats}) / zapped.sats)
+        END) AS weighted_vote,
+        CASE WHEN zapped.sats = 0 THEN 1 ELSE 0 END AS first_vote
+      FROM zapper, zapped
+    )
+    UPDATE "Item"
+    SET
+      "weightedVotes" = "weightedVotes" + zap.weighted_vote,
+      upvotes = upvotes + zap.first_vote,
+      msats = msats + ${msats},
+      "lastZapAt" = now()
+    FROM zap
+    WHERE id = ${itemAct.itemId}`
```

Comment:
@ekzyis you might find this interesting. It's good to be aware of at least. There was a lot of conflicting info about CTEs and subqueries out there, but this kind of thing is unsafe in `read committed`.

Reply:
Resolved in [77cd1a3](https://github.com/stackernews/stacker.news/pull/1195/commits/77cd1a398cb60a05f8365a64777da16492ac1373) by creating a table for storing user<->item aggregates that can receive row level locks.

Reply:
I have read https://www.cybertec-postgresql.com/en/transaction-anomalies-with-select-for-update/ now and it makes sense.

I assume you have decided to go for a table that aggregates data instead of using `REPEATABLE READ` transaction isolation level since that would mean we would need to do retries which is bad for performance? Might also be less safe since we'd have to make sure we always use this isolation level where we could run into such problems. So having a dedicated table for such queries without lock footguns makes sense.